File: bin/strap-runner/env.js
import fs from "node:fs";
import path from "node:path";

const SOURCE = path.dirname(new URL(import.meta.url).pathname);
const DEFAULT_ROOT = path.resolve(SOURCE, "../..");

export function workspaceRoot() {
  return path.resolve(process.env.STRAP_WORKSPACE || process.cwd());
}

export function strapRoot() {
  return path.resolve(process.env.STRAP_ROOT || DEFAULT_ROOT);
}

export function strapGlobalRoot() {
  if (process.env.STRAP_GLOBAL) return path.resolve(process.env.STRAP_GLOBAL);
  return path.join(process.env.XDG_CONFIG_HOME || path.join(process.env.HOME || "", ".config"), "strap");
}

export function strapConfigRoot() {
  if (process.env.STRAP_CONFIG) return path.resolve(process.env.STRAP_CONFIG);
  const repoConfig = path.join(strapRoot(), "config", "strap");
  if (fs.existsSync(repoConfig)) return repoConfig;
  return strapGlobalRoot();
}

function findNearestDir(name, start = process.cwd()) {
  let current = path.resolve(start);
  while (true) {
    const candidate = path.join(current, name);
    if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) return candidate;
    const parent = path.dirname(current);
    if (parent === current) return undefined;
    current = parent;
  }
}

export function strapProjectRoot() {
  if (process.env.STRAP_PROJECT) return path.resolve(process.env.STRAP_PROJECT);
  return findNearestDir(".strap", workspaceRoot()) || path.join(workspaceRoot(), ".strap");
}

export function strapWorkRoot() {
  if (process.env.STRAP_WORK) return path.resolve(process.env.STRAP_WORK);
  return findNearestDir(".strap-user", workspaceRoot()) || path.join(workspaceRoot(), ".strap-user");
}

export function strapSessionRoot() {
  return process.env.STRAP_SESSION ? path.resolve(process.env.STRAP_SESSION) : undefined;
}

export function envFor(name, dir) {
  return {
    ...process.env,
    STRAP_ROOT: strapRoot(),
    STRAP_CONFIG: strapConfigRoot(),
    STRAP_GLOBAL: strapGlobalRoot(),
    STRAP_PROJECT: strapProjectRoot(),
    ...(strapSessionRoot() ? { STRAP_SESSION: strapSessionRoot() } : {}),
    STRAP_WORK: strapWorkRoot(),
    STRAP_WORKSPACE: workspaceRoot(),
    STRAP_CMD_NAME: name,
    STRAP_CMD_DIR: dir,
  };
}

export function resolveSessionEnvFromPointer() {
  if (process.env.STRAP_SESSION) return;
  const current = path.join(strapWorkRoot(), "sessions", "current");
  if (!fs.existsSync(current)) return;
  const session = fs.readFileSync(current, "utf8").trim();
  if (session) process.env.STRAP_SESSION = path.resolve(session);
}

File: subprojects/strap-main/src/strap_main/env.clj
(ns strap-main.env
  (:require [babashka.fs :as fs]
            [clojure.string :as str]))

(def session-override (atom nil))

(defn env [name]
  (System/getenv name))

(defn abs [value]
  (str (fs/absolutize value)))

(defn default-root []
  (-> *file*
      fs/file
      fs/absolutize
      fs/parent
      fs/parent
      fs/parent
      fs/parent
      fs/parent
      str))

(defn workspace-root []
  (abs (or (env "STRAP_WORKSPACE") (System/getProperty "user.dir"))))

(defn strap-root []
  (abs (or (env "STRAP_ROOT") (default-root))))

(defn nearest-dir [name start]
  (loop [dir (fs/absolutize start)]
    (let [candidate (fs/path dir name)
          parent (or (fs/parent dir) dir)]
      (cond
        (fs/directory? candidate) (str candidate)
        (= dir parent) nil
        :else (recur parent)))))

(defn strap-global-root []
  (if-let [root (env "STRAP_GLOBAL")]
    (abs root)
    (str (fs/path (or (env "XDG_CONFIG_HOME")
                      (str (or (env "HOME") "") "/.config"))
                  "strap"))))

(defn strap-project-root []
  (if-let [root (env "STRAP_PROJECT")]
    (abs root)
    (or (nearest-dir ".strap" (workspace-root))
        (str (fs/path (workspace-root) ".strap")))))

(defn strap-work-root []
  (if-let [root (env "STRAP_WORK")]
    (abs root)
    (or (nearest-dir ".strap-user" (workspace-root))
        (str (fs/path (workspace-root) ".strap-user")))))

(defn strap-config-root []
  (if-let [root (env "STRAP_CONFIG")]
    (abs root)
    (let [repo-config (fs/path (strap-root) "config" "strap")]
      (if (fs/exists? repo-config)
        (str repo-config)
        (strap-global-root)))))

(defn strap-session-root []
  (some-> (or @session-override (env "STRAP_SESSION")) abs))

(defn set-session! [session]
  (reset! session-override session))

(defn resolve-session-env-from-pointer []
  (when-not (env "STRAP_SESSION")
    (let [current (fs/path (strap-work-root) "sessions" "current")]
      (when (fs/exists? current)
        (let [session (str/trim (slurp (str current)))]
          (when-not (str/blank? session)
            (set-session! (abs session))))))))

(defn env-for [name dir]
  (cond-> (into {} (System/getenv))
    true (assoc "STRAP_ROOT" (strap-root)
                "STRAP_CONFIG" (strap-config-root)
                "STRAP_GLOBAL" (strap-global-root)
                "STRAP_PROJECT" (strap-project-root)
                "STRAP_WORK" (strap-work-root)
                "STRAP_WORKSPACE" (workspace-root)
                "STRAP_CMD_NAME" name
                "STRAP_CMD_DIR" dir)
    (strap-session-root) (assoc "STRAP_SESSION" (strap-session-root))))
